### PR TITLE
OSASINFRA-3585: openstack: mount config drive when needed

### PR DESCRIPTION
**Pull request by EmilienM**
We need to mount and umount the config drive to get the hostname.

### CODE REVIEW

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 1]
```diff
@@ -31,15 +31,23 @@ contents:
...
-        hostname=$(jq -re .name /host/var/config/openstack/2012-08-10/meta_data.json)
+        hostname=$(jq -re .name $meta_data_file)
         if [[ -z "${hostname}" ]]; then
-            echo "Failed to get hostname from /host/var/config/openstack/2018-08-10/meta_data.json"
+            echo "Failed to get hostname from $meta_data_file"
             exit 1
         fi
 
         echo "KUBELET_NODE_NAME=${hostname}" > ${NODEENV}
+        umount $tmp_mount
```
Comment (MaysaMacedo) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:50]: do we need to unmount? or is it to keep the pattern done in ignition?

Comment (EmilienM) [reply templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:50]: > do we need to unmount? or is it to keep the pattern done in ignition?

yes, cleaner I guess.


>  Also should it be unmounted inside the first condition?

yes i'll add it.

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 2]
```diff
@@ -31,16 +31,33 @@ contents:
     # Set node name to be instance name instead of the default FQDN hostname.
     # Try to get the hostname from config drive first if it's available
     # or fallback to the metadata service.
-    if [ -f "/host/var/config/openstack/2012-08-10/meta_data.json" ]; then
+    if [ ! -z "$(blkid -t LABEL="config-2" -odevice)" ]; then
```
Comment (stephenfin) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:34]: nit:

```suggestion
if [ -n "$(blkid -t LABEL="config-2" -odevice)" ]; then
```

Same thing, no?

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 3]
```diff
@@ -31,16 +31,33 @@ contents:
...
+        mount -o ro -t auto /dev/disk/by-label/$config_drive_label $tmp_mount
+        meta_data_file=$tmp_mount/openstack/2012-08-10/meta_data.json
+        if [ ! -f $meta_data_file ]; then
+            echo "Failed to find $meta_data_file"
+            umount $tmp_mount && rmdir $tmp_mount
+            exit 1
+        fi
         # https://docs.openstack.org/nova/victoria/user/metadata.html#metadata-openstack-format
-        hostname=$(jq -re .name /host/var/config/openstack/2012-08-10/meta_data.json)
+        hostname=$(jq -re .name $meta_data_file)
```
Comment (stephenfin) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:51]:
```suggestion
        hostname=$(jq -re .name "${meta_data_file}")
```

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 4]
```diff
@@ -31,16 +31,33 @@ contents:
...
+    if [ ! -z "${config_drive_label}" ]; then
+        echo "Using config drive to get the hostname"
+        tmp_mount=$(mktemp -d)
+        mount -o ro -t auto /dev/disk/by-label/$config_drive_label $tmp_mount
+        meta_data_file=$tmp_mount/openstack/2012-08-10/meta_data.json
+        if [ ! -f $meta_data_file ]; then
+            echo "Failed to find $meta_data_file"
+            umount $tmp_mount && rmdir $tmp_mount
+            exit 1
+        fi
```
Comment (stephenfin) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:49]:
```suggestion
    if [ ! -z "${config_drive_label}" ]; then
        echo "Using config drive to get the hostname"
        tmp_mount="$(mktemp -d)"
        mount -o ro -t auto "/dev/disk/by-label/${config_drive_label}" "${tmp_mount}"
        meta_data_file="$tmp_mount/openstack/2012-08-10/meta_data.json"
        if [ ! -f "${meta_data_file}" ]; then
            echo "Failed to find ${meta_data_file}"
            umount "${tmp_mount}" && rmdir "${tmp_mount}"
            exit 1
        fi
```

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 5]
```diff
@@ -31,16 +31,33 @@ contents:
...
-        hostname=$(jq -re .name /host/var/config/openstack/2012-08-10/meta_data.json)
+        hostname=$(jq -re .name $meta_data_file)
         if [[ -z "${hostname}" ]]; then
-            echo "Failed to get hostname from /host/var/config/openstack/2018-08-10/meta_data.json"
+            echo "Failed to get hostname from $meta_data_file"
             exit 1
         fi
 
         echo "KUBELET_NODE_NAME=${hostname}" > ${NODEENV}
+        umount $tmp_mount && rmdir $tmp_mount
```
Comment (stephenfin) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:58]:
```suggestion
        umount "${tmp_mount}" && rmdir "${tmp_mount}"
```

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 6]
```diff
@@ -31,16 +31,33 @@ contents:
...
+            echo "Failed to find $meta_data_file"
+            umount $tmp_mount && rmdir $tmp_mount
+            exit 1
+        fi
         # https://docs.openstack.org/nova/victoria/user/metadata.html#metadata-openstack-format
-        hostname=$(jq -re .name /host/var/config/openstack/2012-08-10/meta_data.json)
+        hostname=$(jq -re .name $meta_data_file)
         if [[ -z "${hostname}" ]]; then
-            echo "Failed to get hostname from /host/var/config/openstack/2018-08-10/meta_data.json"
+            echo "Failed to get hostname from $meta_data_file"
```
Comment (stephenfin) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:53]: You've forgotten (?) to unmount the config-drive here.

**Review by stephenfin — CHANGES_REQUESTED** [review]
Mostly nits bar the potential stale mount left around. Feel free to ignore the others if you want. Too much time spent around shellcheck :sweat_smile:

**Comment by EmilienM** [comment]
thanks @stephenfin for the review, it's very appreciated. I addressed all comments.

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 7]
```diff
@@ -31,16 +31,38 @@ contents:
...
     # Try to get the hostname from config drive first if it's available
     # or fallback to the metadata service.
-    if [ -f "/host/var/config/openstack/2012-08-10/meta_data.json" ]; then
+    # We'll first try to use config-drive: if a block device is found with a specific label,
+    # we'll mount it and unmount it when not needed anymore. This is because the config drive is
+    # unmounted by cloud-init.
+    # If config-drive isn't found, then we'll try to reach the metadata server.
+    if [ -n "$(blkid -t LABEL="config-2" -odevice)" ]; then
+        config_drive_label="config-2"
+    elif [ ! -z "$(blkid -t LABEL="CONFIG-2" -odevice)" ]; then
```
Comment (MaysaMacedo) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:40]: Stephen's suggestion could also be applied here. 
```suggestion
    elif [ -n "$(blkid -t LABEL="CONFIG-2" -odevice)" ]; then
```

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml [hunk 8]
```diff
@@ -31,16 +31,38 @@ contents:
...
+        config_drive_label="CONFIG-2"
+    fi
+
+    if [ ! -z "${config_drive_label}" ]; then
```
Comment (MaysaMacedo) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:44]:
```suggestion
    if [ -n "${config_drive_label}" ]; then
```

**Pull request merged**